---
title: "Quantitative Genetics in R"
output: html_document
---

## Introduction

For this module we will be using R and Rstudio. The lesson assumes no prior knowledge of R, but some experience may be helpful to some learners. At the end of the lesson, learners will be able to:

-Open and type commands into R and/or Rstudio, and have them executed in the console

-How to get help with R errors or warnings

-Have an understanding that there are user/community contributions called 'packages' that can help with specific areas

-How to calculate LD for a set of markers and individuals

-How to calculate breeding values

-How to calculate heritability (broad and narrow sense)


# Setup (to be done prior to module)

For this module you will need to download base R and rstudio (if you don't already have them). For this lesson we will be using R 3.4.3 "Kite-Eating Tree". There are distributions of base R for all operating major operating systems.

# **Download R first (to be done prior to module)**

-MAC OS users go here: https://cran.r-project.org/bin/macosx/R-3.4.3.pkg

-Windows users go here: https://cran.r-project.org/bin/windows/base/


# **Download RStudio (to be done prior to module)**
  
Rstudio is an IDE (integrated development environment) for R, and it streamlines writing of scripts, project management, and plotting - click the link below and choose "Rstudio desktop" (free version) and choose your OS version:

https://www.rstudio.com/products/rstudio/download/#download


# **Download Data Set (to be done prior to module)**

For this module we will be using a Picea glauca (white spruce) dataset of 15 phenotypes, 1694 individual trees, and some 6000 SNP markers. The markers have no linkage map, nor does the data set include pedigree based information. Picea glauca like most conifers, has a very large estimated genome size of 16.15 pg or 2100cM (estimated genetic map size), and 12 autosomal chromosomes. This means that we have very low marker coverage - this is further detailed in the paper (Beaulieu et al. 2014). White spruce is widespread in the Boreal and used mostly in residential home construction in Canada and paper making. 

We will use a slightly altered version of the original dataset for the sake of time; however, the original dataset and associated paper by Beaulieu et al. (2014) lives here if anyone is interested in taking a look.

http://datadryad.org/resource/doi:10.5061/dryad.6rd6f


# Part 1: Getting used to and oriented to R

### **Open up RStudio**
Since this is the first time we've opened R let's list the objects in our
workspace:
```{r}
ls()
```

There should be no objects in your workspace. If there are you can wipe the workspace in rstudio using the tiny *broom* icon in the environment panel. Or, you can simply type this command (**NOTE only use this command if you are certain you want to wipe all objects from the workspace**). 

```{r}
rm(list=ls())

ls()
```

You can type these directly in the console, or preferrably in our script file. If we type commands in our console, they will be more difficult to recall or remember.

## Load the data and understand its structure

Now that are workspace is a blank space we can start working with our spruce dataset. First we will load the phenotypes into the workspace:

```{r}
phenos <- read.table("S1.txt", header = TRUE, sep="\t")
```

If you're new to R, you'll notice that when you executed this command on the console - nothing happened. That's generally good news, as it means your command was successfully executed. You'll notice that if you type ```ls()``` an object named ```"phenos"```. 

The first thing to do before examining any data is examine it and visualize it. R has several useful commands to do this. First, let's figure out the size of our data. 

```{r}
dim(phenos)
```

The output of this command tells us that we have an array of 1694 rows x 15 columns. The rows always come first, columns second in R. That's great; however, we would like to get a better picture of the data. We'll do that with the command ```str()``` or the structure command.

```{r}
str(phenos)
```

This not only gives us the dimensions of the data, the datatype we are dealing with, the column names, and the column classes. What does ```str()``` mean though? We can ask R. Type the command below for any function and you'll get the help file to pop up with any arguments that might be contained within that fuction.

```{r}
?str()
```

There's also another function ```summary()``` which produces a brief summary of all data within our ```data.frame()``` columns.

```{r}
summary(phenos)
```

The ```summary()``` can be very useful for figuring out if there's a difference between mean and median values for example. 

## Phenotypic correlations and phenotypic covariance matrix

Since phenotypic correlations are important to consider in multivariate breeders' selection we can produce this matrix in R.

```{r, eval=FALSE}
cor(phenos)
```

An error is generated ```Error in cor(phenos) : x must be numeric```. If you remember from the ```str``` and ```summary``` commands, one of the columns is not numeric.

If we re-execute our command, this time without that column, it should work. We can do this with indexing. 

```{r}
cor(phenos[,-1])
```

We can exclude multiple columns by using the combine function ```c()```

```{r}
cor(phenos[,-c(1:3)])
cor(phenos[,-c(1,4:9,10)])
```

Let's generate the phenotypic covariance matrix, that we could actually use in a breeding program.

```{r}
pheno_covar <- cov(phenos[,-c(1:3)])
pheno_cor <- cor(phenos[,-c(1:3)])
```

## Visualize the data

R comes with numerous helpful commands for plotting and visualizing data. For our phenotypic dataset, let's use a couple basic plots to get a sense of our trait data

```{r}
hist(phenos$cryst)
plot(phenos$coars ~ phenos$cryst)
pairs(pheno_cor)
```

## Using packages 
One of the other positives with using R is that its open-source community contributes many pacakges. A package is a set of functions contributed by a/many users for a specific area of study. Packages may have dependencies on other packages. R has a very active community, and there are many thousands of packages that users can take advantage of, the field of genetics is no exception.

For this lesson, we will use a couple of pkgs ```rrBLUP``` from Jeff Endelman's lab at the University of Wisconsin Madison. We will install the package and load it into the R workspace with the function:

```{r, eval = FALSE}
install.packages("rrBLUP")
library(rrBLUP)
```

If we ever want to read more about a package we can use the ```citation``` function, that is: ```citation(rrBLUP)``` to read more about who generated the package or ```help()```, that is: ```help(rrBLUP)```.

```{r}
citation("rrBLUP")
help(rrBLUP)
```


# Part 2: Quantitative genetics in R

## Load the genotype data

## Compute the additive matrix

## Get an estimate of population structure

## Perform a GWAS 


# Lesson "Easter Eggs" - 

Sometimes it's useful to gather scripts that you can use or modify for the task at hand. A few of these are provided below, but we won't cover them today. You may want to try it out at a later date.

## Recode nucleotide data to a numeric format

Recoding SNP nuleotide data is one of the things that is a common requisite for many packages for quantitative and population genetics. E.g. taking "AA", "CC", "AC", and translating it to 2, 1, 0 or 1,0,-1. There are other formats too. The function below can do this quickly (try it out later if you'd like).

The original spruce dataset from Data Dryad is loaded into the workspace, we define our function, and then we apply it to the ```genos```.

```{r}
genos <- read.table("S2.txt", header = TRUE, sep="\t", stringsAsFactors = TRUE)

genos[1:5,1:5]

recode.SNP.rrBLUP <- function(x) {
  alleles <- unique(x)
  y <- rep(0,length(x))
  y[which(x==alleles[1])] <- -1
  y[which(x==alleles[2])] <- 0
  y[which(x==alleles[3])] <- 1
  y[which(x=="N")] <- NA
  return(y)
}


X <- apply(genos[,-1],2,recode.SNP.rrBLUP)

X[1:5,1:5]
```

If you want the 0, 1, 2 format that is often the required by other programs, you can simply:

```{r}
Xadd <- X + 1

Xadd[1:5,1:5]
```
